### PR TITLE
Partition HTTP Connections

### PR DESCRIPTION
**Pull request by MattMenke2**
Extend the HTTP cache partitioning logic to partition connections as well. This doesn't cover everything that will need to be partitioned, since much of it isn't even capturing in the fetch spec yet (browser-managed DNS caches, WebSocket connections over H2/H3 proxies, H2/H3 sessions, TLS/H3 session resumption cache, cache of metadata about H3-capable servers, HTTP auth cache, etc).

This addresses parts of issue #917.

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * FireFox
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * ...
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/master/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://crbug.com/963480
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1572544
   * Safari: …

***
<a href="https://whatpr.org/fetch/1063.html" title="Last updated on Aug 6, 2020, 2:51 PM UTC (8cbe436)">Preview</a> | <a href="https://whatpr.org/fetch/1063/4ad496d...8cbe436.html" title="Last updated on Aug 6, 2020, 2:51 PM UTC (8cbe436)">Diff</a>

### CODE REVIEW

File: fetch.bs [hunk 1]
```diff
@@ -2071,15 +2071,16 @@ for each associated <a for="fetch group">fetch record</a> whose
 <a for=connection>connection pool</a> consists of zero or more
 <dfn lt=connection export id=concept-connection>connections</dfn>. Each
 <a>connection</a> is identified by an <b>origin</b> (an
-<a for=/>origin</a>) and <b>credentials</b> (a boolean).
+<a for=/>origin</a>), <b>credentials</b> (a boolean), and
+<b>httpStatePartition</b> (a <a for=/>site</a> list or null).
```
Comment (MattMenke2) [inline fetch.bs:2075]: I have no idea how to verify that this "for" link actually works.

**Comment by MattMenke2** [comment]
@annevk @shivanigithub @domenic @mnot thoughts? Not sure if this is the best approach.

In terms of testing, it looks like BasicHttpServer doesn't exposed any sort of connection ID, but it does expose port. So it seems like the best we could do without touching BasicHttpServer itself is to make a request handler that exposes the client-side port each request is using, and make sure the same ports are used to service requests made by two top-level browsing contexts using different origins.  Haven't looked into the exact mechanics of how to share observed port data, but could just use link decoration with a list of previously observed ports, if there's nothing more elegant available.

**Comment by MattMenke2** [comment]
I have no idea how to address those "make deploy" errors ("file:/home/runner/work/fetch/fetch/fetch.spec.whatwg.org/index.html":1958.5-1958.36: error: Element “aside” not allowed as child of element “dl” in this context. (Suppressing further errors from this subtree.))

I don't have access to index.html, so have no idea what's on line 1958. Running make deploy locally, it complains about GITHUB_REPOSITORY not existing, and setting it manually doesn't seem to address that. Looks like that's just for lines htat don't run, regardless, and the error appears to be on the step that's gated on the GITHUB_REPOSITORY step. Also, while running make deploy locally claims that it's generating index.html (".spec.whatwg.org/index.html"), I still don't have any idea how to access that file, to see what's at line 1958.

File: fetch.bs [hunk 2]
```diff
@@ -4531,12 +4536,15 @@ Range Requests</cite>. [[HTTP-RANGE]] However, this is not widely supported by b
     <var>httpRequest</var>'s
     <a for=request>credentials mode</a>.
 
-   <li><p>Set <var>httpCache</var> to the result of <a>determining the HTTP cache partition</a>,
+   <li><p>Set <var>httpStatePartition</var> to the result of <a>determining the HTTP state partition</a>,
    given <var>httpRequest</var>.
 
    <li><p>If <var>httpCache</var> is null, then set <var>httpRequest</var>'s
```
Comment (shivanigithub) [inline fetch.bs:4542]: Since the httpCache is replaced by httpStatePartition above, <var>httpCache</var> here will not be a valid cache?

**Comment by shivanigithub** [comment]
Regarding the deploy errors, I don't see an <aside> element in your changes which makes me wonder if the error exists without your changes as well?

**Comment by shivanigithub** [comment]
Regarding the deploy errors, I don't see an \<aside\> element in your changes which makes me wonder if the error exists without your changes as well?

**Comment by domenic** [comment]
Sorry about the deploy errors. That is a problem with the generated code via the build tool, and will be fixed by https://github.com/tabatkins/bikeshed/pull/1727.

File: fetch.bs (hunk 1 above)
Comment (shivanigithub) [inline fetch.bs:2075]: Does this mean that an httpStatePartition is a list of sites? May be add "key" to the name to be more clear that this is not the actual partition but the key "httpStatePartitionKey"?

File: fetch.bs [hunk 3]
```diff
@@ -4796,7 +4804,7 @@ Range Requests</cite>. [[HTTP-RANGE]] However, this is not widely supported by b
 
  <li><p>Let <var>response</var> be null.
 
- <li><p>Let <var>httpCache</var> be the result of <a>determining the HTTP cache partition</a>, given
+ <li><p>Let <var>httpCache</var> be the result of <a>determining the HTTP state partition</a>, given
```
Comment (shivanigithub) [inline fetch.bs:4807]: Let <var>httpCache</var> be the HTTP cache associated with the result of <a>determining the HTTP cache partition</a>

File: fetch.bs (hunk 2 above)
Comment (MattMenke2) [reply fetch.bs:4547]: Thanks! Meant to put the line below this above it, but even that doesn't really handle the null case. Fixed.

File: fetch.bs (hunk 3 above)
Comment (MattMenke2) [reply fetch.bs]: Thanks! Meant to put the line below this above it, but even that doesn't really handle the null case. Fixed.

File: fetch.bs (hunk 1 above)
Comment (MattMenke2) [reply fetch.bs]: Done, everywhere (also "HTTP state partition" -> "HTTP state partition key")

File: fetch.bs [hunk 4]
```diff
@@ -2123,11 +2125,11 @@ clearly stipulates that <a>connections</a> are keyed on
      WebSocket saner -->
 
 
-<h3 id=http-cache-partitions>HTTP cache partitions</h3>
+<h3 id=http-state-partition>HTTP state partition key</h3>
```
Comment (annevk) [inline fetch.bs:2128]: Should we make this network or perhaps fetch partition key? Network might be better as we'll want to use it for things that do not go through Fetch as well, such as RTCDataChannel, right?

File: fetch.bs [hunk 5]
```diff
@@ -4531,9 +4536,14 @@ Range Requests</cite>. [[HTTP-RANGE]] However, this is not widely supported by b
     <var>httpRequest</var>'s
     <a for=request>credentials mode</a>.
 
-   <li><p>Set <var>httpCache</var> to the result of <a>determining the HTTP cache partition</a>,
+   <li><p>Set <var>httpStatePartitionKey</var> to the result of <a>determining the HTTP state partition key</a>,
    given <var>httpRequest</var>.
 
+   <li><p>Set <var>httpCache</var> to null.
```
Comment (annevk) [inline fetch.bs:4542]: How can it be non-null at this point?

File: fetch.bs [hunk 6]
```diff
@@ -4796,8 +4806,8 @@ Range Requests</cite>. [[HTTP-RANGE]] However, this is not widely supported by b
 
  <li><p>Let <var>response</var> be null.
 
- <li><p>Let <var>httpCache</var> be the result of <a>determining the HTTP cache partition</a>, given
- <var>httpRequest</var>.
+ <li><p>Let <var>httpCache</var> be the result of <a>determining the HTTP state partition key</a>,
+ given <var>httpRequest</var>.
```
Comment (annevk) [inline fetch.bs:4810]: This returns a key, not a cache. I think that this illustrates that we still want an easy way to obtain a partitioned cache though and not inline that in various places.

File: fetch.bs (hunk 5 above)
Comment (MattMenke2) [reply fetch.bs]: I assume we need to declare variables exist before conditionally setting them to something.  If this were psuedo-Javascript, it's undefined at this point, not null. If this is psuedo-C, it's a compile failure.

Anyhow, I've reverted this and added back a shortened HTTP cache partition section, per your other comment.

File: fetch.bs (hunk 6 above)
Comment (MattMenke2) [reply fetch.bs]: This would be just as accurate if this were null-checking checking the network partition key, and I just renamed httpCache here to networkPartitionKey - I just forgot to rename the variable.  That having been said, I've brought back the cache section, though it's pretty tiny now, and reverted this to directly reference the cache. That does seem less regression prone.

File: fetch.bs (hunk 5 above)
Comment (annevk) [reply fetch.bs]: There's a much earlier "Let _httpCache_ be null." step, no?

**Comment by annevk** [comment]
@MattMenke2 you'll need to make your membership of the googlers organization public. Hopefully once https://github.com/tabatkins/bikeshed/pull/1742 is fixed the build issues will go away too.

File: fetch.bs [hunk 7]
```diff
@@ -2123,11 +2125,11 @@ clearly stipulates that <a>connections</a> are keyed on
      WebSocket saner -->
 
 
-<h3 id=http-cache-partitions>HTTP cache partitions</h3>
+<h3 id=http-state-partition>Network partition key</h3>
```
Comment (annevk) [inline fetch.bs:2129]: Let's add a definition here of a network partition key as a tuple consisting of a site and null or an implementation-defined value.

File: fetch.bs [hunk 8]
```diff
@@ -2158,12 +2160,31 @@ given <var>request</var>, run these steps:
  <li><p>Let <var>topLevelSite</var> be the result of <a lt="obtain a site">obtaining a site</a>,
  given <var>topLevelOrigin</var>.
 
- <li>
-  <p>Return the HTTP cache associated with <var>topLevelSite</var> and, possibly, a second key.
-  [[!HTTP-CACHING]]
+ <li><p>Let <var>secondKey</var> be the second key, or null.
```
Comment (annevk) [inline fetch.bs:2163]: Let's rephrase this to

> `be null or an <a>implementation-defined</a> value.`

to make it even more explicit.

File: fetch.bs [hunk 9]
```diff
@@ -2158,12 +2160,31 @@ given <var>request</var>, run these steps:
...
 
   <p class=XXX>The second key is intentionally a little vague as the finer points are still
   evolving. See <a href=https://github.com/whatwg/fetch/issues/1035>issue #1035</a>.
+
+ <li><p>If <var>secondKey</var> is not null, return (<var>topLevelSite</var>, <var>secondKey</var>).
```
Comment (annevk) [inline fetch.bs:2168]: Let's remove the conditional and the then unneeded step below. Implementations that do not have a second key can use a simpler data structure.

File: fetch.bs [hunk 10]
```diff
@@ -2071,15 +2071,16 @@ for each associated <a for="fetch group">fetch record</a> whose
 <a for=connection>connection pool</a> consists of zero or more
 <dfn lt=connection export id=concept-connection>connections</dfn>. Each
 <a>connection</a> is identified by an <b>origin</b> (an
-<a for=/>origin</a>) and <b>credentials</b> (a boolean).
+<a for=/>origin</a>), <b>credentials</b> (a boolean), and
+<b>networkPartitionKey</b> (a <a for=/>site</a> list or null).
```
Comment (annevk) [inline fetch.bs:2075]: I suggest we call this key and lead with it. So key, origin, credentials. And per below its type can be network partition key.

File: fetch.bs [hunk 11]
```diff
@@ -2123,11 +2125,11 @@ clearly stipulates that <a>connections</a> are keyed on
      WebSocket saner -->
 
 
-<h3 id=http-cache-partitions>HTTP cache partitions</h3>
+<h3 id=http-state-partition>Network partition key</h3>
```
Comment (annevk) [inline fetch.bs:2128]: The ID here needs updating and the title should be plural per convention.

File: fetch.bs [hunk 12]
```diff
@@ -2158,12 +2160,31 @@ given <var>request</var>, run these steps:
...
+
+
+<h3 id=http-cache-partitions>HTTP cache partitions</h3>
+
+<p>To
+<dfn lt="determine the HTTP cache partition|determining the HTTP cache partition">determine the HTTP cache partition</dfn>,
+given <var>request</var>, run these steps:
+
+<ol>
+ <li><p>Set <var>networkPartitionKey</var> to the result of <a>determining the network partition key</a>.
```
Comment (annevk) [inline fetch.bs:2182]: Use "Let" when initializing. See https://infra.spec.whatwg.org/#variables.

File: fetch.bs [hunk 13]
```diff
@@ -2158,12 +2160,31 @@ given <var>request</var>, run these steps:
...
+
+ <li><p>If <var>networkPartitionKey</var> is non-null, return null.
```
Comment (annevk) [inline fetch.bs:2184]: then return null.

But also, the conditional seems wrong?

File: fetch.bs [hunk 14]
```diff
@@ -2158,12 +2160,31 @@ given <var>request</var>, run these steps:
...
+
+ <li><p>Otherwise, Return the unique HTTP cache associated with the <var>networkPartitionKey</var>.
```
Comment (annevk) [inline fetch.bs:2186]: No need for otherwise.

File: fetch.bs (hunk 10 above)
Comment (MattMenke2) [reply fetch.bs]: Done. Renamed it in HTTP cache partitions section as well, but stuck with networkPartitionKey in the logic to find a connection, since that's a longer method.

File: fetch.bs (hunk 7 above)
Comment (MattMenke2) [reply fetch.bs:2131]: Done.

File: fetch.bs (hunk 9 above)
Comment (MattMenke2) [reply fetch.bs]: Done.

File: fetch.bs (hunk 5 above)
Comment (MattMenke2) [reply fetch.bs]: You're correct - completely missed it, given how far up it is.

**Comment by MattMenke2** [comment]
> @MattMenke2 you'll need to make your membership of the googlers organization public. Hopefully once [tabatkins/bikeshed#1742](https://github.com/tabatkins/bikeshed/pull/1742) is fixed the build issues will go away too.

My membership of the googlers organization isn't listed. Not sure how to resolve that.

**Comment by MattMenke2** [comment]
> > 
> 
> My membership of the googlers organization isn't listed. Not sure how to resolve that.

Erm...To be clear, googlers doesn't appear in my organization list.

**Comment by MattMenke2** [comment]
> > > 
> > 
> > 
> > My membership of the googlers organization isn't listed. Not sure how to resolve that.
> 
> Erm...To be clear, googlers doesn't appear in my organization list.

This should be taken care of (Thanks Domenic!).   It is rather unfortunate that this required me to hand over my personal phone number to github (I don't have a work phone, since I don't really need one).  :(

**Comment by shivanigithub** [comment]
Looks good to me as well.

**Comment by MattMenke2** [comment]
@annevk: Your changes all look good to me.

**Comment by annevk** [comment]
Thanks all! (To be clear, I'm not blocking on tests as we don't have good infrastructure to test connections and I don't really see that changing anytime soon.)

**Comment by MattMenke2** [comment]
> Thanks all! (To be clear, I'm not blocking on tests as we don't have good infrastructure to test connections and I don't really see that changing anytime soon.)

Thanks again, Anne! I do have some putative tests that I hope to send out this week or next. They're sufficiently weird that it's possible I'll face pushback on landing them, though. They track which tabs send requests to which server ports, with the server code assuming ports aren't reused, since the Python server class only exposes port information, not connection information, and doesn't expose when connections are closed.  I'll plan to go through review github instead of Chrome's magic mirror, to allow a broader set of folks to offer feedback or push back on the changes.

**Comment by annevk** [comment]
Oh, that sounds great. I'll be out for week but happy to look at that mid-August.
